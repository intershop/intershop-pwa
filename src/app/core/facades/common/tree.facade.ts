import { InjectionToken } from '@angular/core';
import { Observable } from 'rxjs';

import { DynamicFlatNode } from 'ish-core/utils/tree/tree.interface';

export const TREE_FACADE_IMPLEMENTOR = new InjectionToken<TreeFacade>('TreeFacadeImplementor');

export interface TreeFacade {
  initialData$(): Observable<DynamicFlatNode[]>;
  getChildren$(node: string): Observable<Omit<DynamicFlatNode, 'level'>[]>;
}
