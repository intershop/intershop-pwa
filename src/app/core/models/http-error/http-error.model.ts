export interface HttpHeader {
  [key: string]: string;
}

export interface HttpError extends Error {
  message: string;
  error: string;
  errorCode: string;
  status: number;
  statusText: string;
  headers: HttpHeader;
}
