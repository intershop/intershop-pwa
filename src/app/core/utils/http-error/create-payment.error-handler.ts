import { SpecialHttpErrorHandler } from 'ish-core/interceptors/icm-error-mapper.interceptor';

export const createPaymentErrorHandler: SpecialHttpErrorHandler = {
  test: (error, request) => error.url.endsWith('/payments') && request.method === 'POST',
  map: error => {
    if (error?.error && error.error.includes('{') && error.error.includes('}')) {
      return { code: error.error.slice(error.error.indexOf('{') + 1, error.error.lastIndexOf('}')) };
    }
  },
};
