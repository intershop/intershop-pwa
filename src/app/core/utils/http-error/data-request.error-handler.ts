import { SpecialHttpErrorHandler } from 'ish-core/interceptors/icm-error-mapper.interceptor';

export const dataRequestErrorHandler: SpecialHttpErrorHandler = {
  test: (error, request) => error.url.endsWith('/confirmations') && request.method === 'PUT',
  map: error => {
    switch (error.status) {
      case 404:
        return { code: 'personal.data.request.confirmation_link_expired.error' };
      case 422:
        return { code: 'personal.data.request.unprocessable.error' };
      default:
        return { code: 'personal.data.request.server_connection_failed.error' };
    }
  },
};
