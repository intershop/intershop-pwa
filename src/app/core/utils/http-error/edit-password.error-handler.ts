import { SpecialHttpErrorHandler } from 'ish-core/interceptors/icm-error-mapper.interceptor';

export const editPasswordErrorHandler: SpecialHttpErrorHandler = {
  test: error =>
    error.url.endsWith('/credentials/password') &&
    error.headers?.get('error-invalid-attributes')?.includes('currentPassword'),
  map: () => ({ code: 'account.update_password.old_password.error.incorrect' }),
};
