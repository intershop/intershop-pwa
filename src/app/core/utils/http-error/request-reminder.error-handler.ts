import { SpecialHttpErrorHandler } from 'ish-core/interceptors/icm-error-mapper.interceptor';

export const requestReminderErrorHandler: SpecialHttpErrorHandler = {
  test: error => error.url.endsWith('/security/reminder'),
  map: error => {
    switch (error.status) {
      case 401:
        return { code: 'captcha.incorrect' };
      case 500:
        return { code: 'servererror.mailServer.error' };
      default:
        return { code: 'account.forgotdata.email.invalid_login_or_name.error' };
    }
  },
};
