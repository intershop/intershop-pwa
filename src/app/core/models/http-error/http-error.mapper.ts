import { HttpErrorResponse } from '@angular/common/http';

import { BasketInfo } from 'ish-core/models/basket-info/basket-info.model';

import { HttpError } from './http-error.model';

export class HttpErrorMapper {
  // tslint:disable-next-line:ban-types
  static fromError(error: HttpErrorResponse): HttpError {
    const headers = !error.headers
      ? undefined
      : error.headers.keys().reduce((acc, val) => ({ ...acc, [val]: error.headers.get(val) }), {});

    return {
      name: error.name,
      message: error.message,
      errorCode: error?.error?.errors?.[0]?.code,
      error:
        typeof error.error === 'string'
          ? error.error
          : HttpErrorMapper.determineErrorMessage(error?.error?.errors?.[0]),
      status: error.status,
      statusText: error.statusText,
      headers,
    };
  }

  private static determineErrorMessage(error: BasketInfo): string {
    let message = error && error.message;
    if (message && error.causes) {
      message += error.causes.map(cause => (cause.message ? ' ' + cause.message : ''));
    }
    return message || undefined; // undefined is needed to avoid null
  }
}
