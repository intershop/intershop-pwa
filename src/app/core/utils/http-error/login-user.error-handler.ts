import { HttpErrorResponse, HttpRequest } from '@angular/common/http';
import { Inject, Injectable } from '@angular/core';

import { USER_REGISTRATION_LOGIN_TYPE } from 'ish-core/configurations/injection-keys';
import { SpecialHttpErrorHandler } from 'ish-core/interceptors/icm-error-mapper.interceptor';
import { HttpError } from 'ish-core/models/http-error/http-error.model';
import { ApiService } from 'ish-core/services/api/api.service';

// tslint:disable: ban-types

@Injectable()
export class LoginUserErrorHandler implements SpecialHttpErrorHandler {
  constructor(@Inject(USER_REGISTRATION_LOGIN_TYPE) public loginType: string) {}

  test(error: HttpErrorResponse, request: HttpRequest<unknown>): boolean {
    return (
      request.headers.has(ApiService.AUTHORIZATION_HEADER_KEY) &&
      error.status === 401 &&
      error.url.includes('customers/-')
    );
  }
  map(): Partial<HttpError> {
    if (this.loginType === 'email') {
      return { code: 'account.login.email_password.error.invalid' };
    } else {
      return { code: 'account.login.username_password.error.invalid' };
    }
  }
}
