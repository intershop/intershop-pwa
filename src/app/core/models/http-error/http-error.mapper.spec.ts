import { HttpErrorResponse, HttpHeaders } from '@angular/common/http';

import { HttpErrorMapper } from './http-error.mapper';

describe('Http Error Mapper', () => {
  it('should convert correctly for simple HttpErrorResponse', () => {
    expect(HttpErrorMapper.fromError(new HttpErrorResponse({ status: 401, statusText: 'Unauthorized' })))
      .toMatchInlineSnapshot(`
      Object {
        "error": undefined,
        "errorCode": undefined,
        "headers": Object {},
        "message": "Http failure response for (unknown url): 401 Unauthorized",
        "name": "HttpErrorResponse",
        "status": 401,
        "statusText": "Unauthorized",
      }
    `);
  });
  it('should convert correctly for HttpErrorResponse with headers', () => {
    expect(
      HttpErrorMapper.fromError(
        new HttpErrorResponse({
          status: 500,
          headers: new HttpHeaders().set('key1', 'value1').set('key2', 'value2'),
        })
      )
    ).toMatchInlineSnapshot(`
      Object {
        "error": undefined,
        "errorCode": undefined,
        "headers": Object {
          "key1": "value1",
          "key2": "value2",
        },
        "message": "Http failure response for (unknown url): 500 undefined",
        "name": "HttpErrorResponse",
        "status": 500,
        "statusText": "Unknown Error",
      }
    `);
  });

  it('should convert correctly for simple HttpErrorResponse with error Message', () => {
    expect(
      HttpErrorMapper.fromError(
        new HttpErrorResponse({ status: 401, statusText: 'Unauthorized', error: 'Error Message' })
      )
    ).toMatchSnapshot();
  });

  it('should convert correctly for HttpErrorResponse with errors and causes', () => {
    expect(
      HttpErrorMapper.fromError(
        new HttpErrorResponse({
          status: 422,
          statusText: 'Unprocessable Entity',
          error: {
            errors: [
              {
                message: 'The product could not be added to your basket.',
                causes: [{ message: 'The maximum allowed item quantity is exceeded' }],
              },
            ],
          },
        })
      )
    ).toMatchSnapshot();
  });
});
