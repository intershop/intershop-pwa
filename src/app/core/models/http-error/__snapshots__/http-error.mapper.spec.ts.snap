// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Http Error Mapper should convert correctly for HttpErrorResponse with errors and causes 1`] = `
Object {
  "error": "The product could not be added to your basket. The maximum allowed item quantity is exceeded",
  "errorCode": undefined,
  "headers": Object {},
  "message": "Http failure response for (unknown url): 422 Unprocessable Entity",
  "name": "HttpErrorResponse",
  "status": 422,
  "statusText": "Unprocessable Entity",
}
`;

exports[`Http Error Mapper should convert correctly for simple HttpErrorResponse with error Message 1`] = `
Object {
  "error": "Error Message",
  "errorCode": undefined,
  "headers": Object {},
  "message": "Http failure response for (unknown url): 401 Unauthorized",
  "name": "HttpErrorResponse",
  "status": 401,
  "statusText": "Unauthorized",
}
`;
